File: solutions/fsharp/proverb/Proverb.fs
module Proverb

let recite (input: string list) : string list =
    let inputCycle : string list =
        match input with
        | [] -> []
        | [ one ] -> [ one ]
        | _ -> List.append input [ (List.head input) ]

    let rec reciteCycle (i: string list) (acc: string list) : string list =
        match i with
        | one :: two :: rest when not rest.IsEmpty ->
            reciteCycle (two :: rest) ($"For want of a {one} the {two} was lost." :: acc)
        | _ :: two :: [] -> $"And all for the want of a {two}." :: acc
        | one :: [] -> $"And all for the want of a {one}." :: acc
        | [] -> []

    reciteCycle inputCycle [] |> List.rev

File: solutions/elixir/proverb/lib/proverb.ex
defmodule Proverb do
  @doc """
  Generate a proverb from a list of strings.
  """
  @spec recite(strings :: [String.t()]) :: String.t()
  def recite([]), do: ""
  def recite([first | _] = strings) do
    pairs(strings) ++ [final_line(first)]
    |> Enum.join("\n")
  end

  defp pairs([_]), do: []
  defp pairs([first, second | rest]) do
    ["For want of a #{first} the #{second} was lost." | pairs([second | rest])]
  end

  defp final_line(string), do: "And all for the want of a #{string}.\n"
end
